### PR TITLE
ORT Seq2Seq modeling refactorization

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
This PR:
- Creates and separates the testing of seq2seq / encoder-decoder models (text, speech and vision).
- Enables full testing, including batched inference with padded inputs and logits, last hidden state and pkv matching assertions. in addition to beam search, random sampling...
- Fixes issues emerging/detected when running the full testing suite, like missing dynamic axis, missing attention masks in the encoder/decoder, untested model types for some tasks, non-traceable methods, etc..
- Rewrites and simplifies the seq2seq modeling code, removing a lot of hackiness around pkv (self attn vs cross attn) and making the forward pass (especially the seq2seqdecoder forward) more linear, finally enabling cases like use_merged=True and use_cache=False, which weren't supported before. 
- Documents behaviors like m2m100 consuming the RNG state and bigbird_pegasus switching between attention types.
- Enables Moonshine speech recognition (inference) and patch it for dynamic audio input length (only supported static).
- Removes some deprecated stuff, like onnx config with loss which was only relevant for ort training.
- Rewrite and simplifies the filename inference when loading models.
- Correct the output dynamic axis for different ctc models.
- [maybe more 🥲]

### CODE REVIEW

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 1]
```diff
@@ -829,6 +829,16 @@ def _cached_file(
             )
         return cached_path
 
+    # TODO: move it to ORTModel (or ORTHubMixin when OptimizedModel is removed)
+    @staticmethod
+    def _infer_file_path(pattern: str, file_name: str, onnx_files: list[Path]):
+        pattern_files = [p for p in onnx_files if re.search(pattern, str(p))]
```
Comment (justinchuby) [inline optimum/onnxruntime/modeling_seq2seq.py]: Is there a naming rule that can be documented?

Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_seq2seq.py]: I'm not aware of any naming rules here, this seemed to me like a repeated pattern of inferring the onnx file path from a pattern with a user given file name, Ideally I want all these file/hub methods to be in ORTHubMixin (or ONNXFileMixin if it makes more sense) but I have already changed too many things in this PR 😅

File: optimum/exporters/onnx/__init__.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
 
 _import_structure = {
-    "base": ["OnnxConfig", "OnnxConfigWithLoss", "OnnxConfigWithPast", "OnnxSeq2SeqConfigWithPast"],
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/__init__.py:23]: the OnnxConfigWithLoss class was being used with ort training

File: tests/onnxruntime/testing_utils.py [hunk 3]
```diff
@@ -49,16 +49,14 @@
...
     "dinov2": "hf-internal-testing/tiny-random-Dinov2Model",
     "distilbert": "hf-internal-testing/tiny-random-DistilBertModel",
     "dpt": "hf-internal-testing/tiny-random-DPTForSemanticSegmentation",
     "electra": "hf-internal-testing/tiny-random-ElectraModel",
-    "encoder-decoder": {
-        "hf-internal-testing/tiny-random-EncoderDecoderModel-bert-bert": ["text2text-generation"],
-        "mohitsha/tiny-random-testing-bert2gpt2": ["text2text-generation", "text2text-generation-with-past"],
-    },
+    "encoder-decoder": "optimum-internal-testing/tiny-random-encoder-decoder-gpt2-bert",
+    "encoder-decoder-bert-bert": "hf-internal-testing/tiny-random-EncoderDecoderModel-bert-bert",
```
Comment (IlyasMoutawwakil) [inline tests/onnxruntime/testing_utils.py:61]: this way the test doesn't need to loop over multiple tests in the same test scope.

File: optimum/exporters/onnx/config.py [hunk 4]
```diff
@@ -171,35 +165,19 @@ class TextSeq2SeqOnnxConfig(OnnxSeq2SeqConfigWithPast):
...
 
-    @property
-    def torch_to_onnx_input_map(self) -> dict[str, str]:
-        if self._behavior is ConfigBehavior.DECODER:
-            return {
-                "decoder_input_ids": "input_ids",
-                "encoder_outputs": "encoder_hidden_states",
-                "attention_mask": "encoder_attention_mask",
-            }
-        return {}
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/config.py:182]: moved to seq2seq config

File: optimum/exporters/onnx/model_configs.py [hunk 5]
```diff
@@ -647,6 +621,8 @@ class LongT5OnnxConfig(T5OnnxConfig):
     *["feature-extraction", "feature-extraction-with-past", "text2text-generation", "text2text-generation-with-past"],
 )
 class M2M100OnnxConfig(TextSeq2SeqOnnxConfig):
+    PAD_ATTENTION_MASK_TO_PAST = True
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:692]: what's padded in this case is the "decoder_attention_mask"

File: optimum/exporters/onnx/model_configs.py [hunk 6]
```diff
@@ -1052,7 +1016,7 @@ def torch_to_onnx_input_map(self) -> dict[str, str]:
         return {"x": "pixel_values"}
 
 
-@register_tasks_manager_onnx("mgp-str", *["feature-extraction", "image-to-text"])
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:1123]: the model doesn't support image to text but rather [scene text recognition](https://huggingface.co/docs/transformers/en/model_doc/mgp-str#transformers.MgpstrForSceneTextRecognition)

File: optimum/exporters/onnx/model_configs.py [hunk 7]
```diff
@@ -1672,6 +1612,34 @@ def generate_dummy_inputs(self, framework: str = "pt", **kwargs):
...
+
+        # Hubert output formula adapted from:
+        # https://github.com/huggingface/transformers/blob/v4.55.2/src/transformers/models/hubert/modeling_hubert.py#L721
+        if self.task == "automatic-speech-recognition":
+            sequence_length = "sequence_length"
+            for kernel_size, stride in zip(self._config.conv_kernel, self._config.conv_stride):
+                sequence_length = f"( {sequence_length} - {kernel_size} ) // {stride} + 1"
+            outputs["logits"] = {0: "batch_size", 1: sequence_length}
+
+        return outputs
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/model_configs.py:1696]: this is valid for most ctc models (except mctct)

File: tests/onnxruntime/test_seq2seq.py [hunk 8]
```diff
@@ -0,0 +1,1343 @@
...
+        onnx_model = self.ORTMODEL_CLASS.from_pretrained(
+            self.onnx_model_dirs[test_name], use_cache=use_cache, use_merged=use_merged, use_io_binding=use_io_binding
+        )
+        return onnx_model
+
+    # INTEGRATION TESTS
+    # The task automatic-speech-recognition contains models like hubert, mctct, sew, etc. that are not supported by the
+    # ORTForSpeechSeq2Seq class, but rather by the ORTModelForCTC class.
+    # def test_find_untested_architectures(self):
+    #     self._test_find_untested_architectures()
```
Comment (IlyasMoutawwakil) [inline tests/onnxruntime/test_seq2seq.py:980]: ctc models are also registered as "automatic-speech-recognition" models

File: optimum/exporters/onnx/config.py [hunk 9]
```diff
@@ -22,6 +22,8 @@
 
 from optimum.exporters.onnx.base import ConfigBehavior, OnnxConfig, OnnxConfigWithPast, OnnxSeq2SeqConfigWithPast
 from optimum.exporters.onnx.constants import ONNX_DECODER_MERGED_NAME, ONNX_DECODER_NAME, ONNX_DECODER_WITH_PAST_NAME
+from optimum.exporters.tasks import TasksManager
+from optimum.onnx import merge_decoders
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/onnx/config.py:26]: this moves back onnx imports, might break optimum-intel on windows ?

File: tests/onnxruntime/test_seq2seq.py [hunk 10]
```diff
@@ -0,0 +1,1328 @@
...
+        self.assertIn("generated_text", outputs[0][0])
+        self.assertIsInstance(outputs[0][0]["generated_text"], str)
+        self.assertGreater(len(outputs[0][0]["generated_text"]), 0)
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            pipe.save_pretrained(tmpdir)
+
+            # TODO: should be fixed in optimum
+            # The pipeline for image-to-text generation fails to load the tokenizer
+            return
```
Comment (IlyasMoutawwakil) [inline tests/onnxruntime/test_seq2seq.py]: fixed in #41

File: tests/onnxruntime/test_seq2seq.py [hunk 11]
```diff
@@ -0,0 +1,1328 @@
...
+        )
+
+    # PIPELINE TESTS
+    @parameterized.expand(grid_parameters({"use_cache": [True], "use_merged": [False, True]}))
+    def test_pipeline_with_default_model(self, test_name: str, use_cache: bool, use_merged: bool):
+        images = self.get_inputs("vision-encoder-decoder", for_pipeline=True)
+
+        # TODO: should be fixed in optimum
+        # The pipeline for image-to-text generation fails to load the tokenizer
+        return
```
Comment (IlyasMoutawwakil) [inline tests/onnxruntime/test_seq2seq.py]: fixed in #41

File: pyproject.toml [hunk 12]
```diff
@@ -29,7 +29,7 @@ classifiers = [
     "Topic :: Scientific/Engineering :: Artificial Intelligence",
 ]
 dependencies = [
-    "optimum @ git+https://github.com/huggingface/optimum",
+    "optimum@git+https://github.com/huggingface/optimum.git@cleanup-tasks",
```
Comment (IlyasMoutawwakil) [inline pyproject.toml]:
```suggestion
    "optimum @ git+https://github.com/huggingface/optimum.git",
```

File: pyproject.toml [hunk 13]
```diff
@@ -29,7 +29,7 @@ classifiers = [
     "Topic :: Scientific/Engineering :: Artificial Intelligence",
 ]
 dependencies = [
-    "optimum @ git+https://github.com/huggingface/optimum",
+    "optimum @ git+https://github.com/huggingface/optimum.git",
```
Comment (IlyasMoutawwakil) [inline pyproject.toml]:
```suggestion
    "optimum @ git+https://github.com/huggingface/optimum",
```

File: pyproject.toml [hunk 14]
```diff
@@ -56,7 +56,7 @@ tests = [
     # not sure they are necessary
     "rjieba",
     "sacremoses",
-    "onnxslim>=0.1.60",
+    "onnxslim>=0.1.60,<0.1.66",
```
Comment (IlyasMoutawwakil) [inline pyproject.toml]:
```suggestion
    "onnxslim>=0.1.60",
```

File: optimum/exporters/onnx/input_generators.py [hunk 15]
```diff
@@ -64,3 +69,27 @@ def generate(self, input_name: str, framework: str = "pt", int_dtype: str = "int
...
+        if input_name == "input_values":  # raw waveform
+            return self.random_float_tensor(
+                shape=[self.batch_size, self.sequence_length],
+                min_value=-1,
+                max_value=1,
+                framework=framework,
+                dtype=float_dtype,
+            )
+        elif input_name == "attention_mask":  # attention mask
+            return self.random_int_tensor(
```
Comment (echarlaix) [inline optimum/exporters/onnx/input_generators.py]: should we use `random_mask_tensor` here ?

File: optimum/exporters/onnx/model_configs.py [hunk 16]
```diff
@@ -2581,118 +2455,52 @@ class Pix2StructOnnxConfig(OnnxSeq2SeqConfigWithPast):
         DummyPix2StructInputGenerator,
     )
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        if is_transformers_version("==", "4.46.0") and self._behavior is ConfigBehavior.DECODER:
```
Comment (echarlaix) [inline optimum/exporters/onnx/model_configs.py:2601]: this is not needed anymore ?

File: optimum/onnxruntime/modeling_ort.py [hunk 17]
```diff
@@ -572,7 +532,7 @@ def can_generate(self) -> bool:
         """Returns whether this model can generate sequences with `.generate()`."""
         return isinstance(self, GenerationMixin)
 
-    def _warn_on_unhandled_inputs(self, kwargs: dict[str, Any]) -> None:
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_ort.py]: would keep it this way (to leave us the room to do any breaking change)

File: optimum/onnxruntime/modeling_ort.py [hunk 18]
```diff
@@ -1611,32 +1553,30 @@ class ORTModelForCTC(ORTModel):
...
         input_values: torch.Tensor | np.ndarray | None = None,
+        input_features: torch.Tensor | np.ndarray | None = None,
         *,
         return_dict: bool = True,
         **kwargs,
     ):
-        # Warn about any unexpected kwargs using the helper method
-        self._warn_on_unhandled_inputs(kwargs)
+        if self.config.model_type == "mctct":
+            assert input_features is not None, "input_features must be provided for this model"
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_ort.py:1606]: would raise an exception instead

File: optimum/onnxruntime/modeling_decoder.py [hunk 19]
```diff
@@ -441,8 +395,8 @@ def forward(
                 self.session.run_with_iobinding(self._io_binding)
                 self._io_binding.synchronize_outputs()
 
-            loss = output_buffers.get("loss", None)
-            logits = output_buffers["logits"].view(output_shapes["logits"])
+            loss = output_buffers.pop("loss", None)
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_decoder.py]: shouldn't we remove the loss here ? (wondering when this would be needed)

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 20]
```diff
@@ -831,7 +587,175 @@ def forward(
...
+
+        if self.config.model_type == "pix2struct":
+            self.vocab_size = getattr(self.config, "text_config", self.config).vocab_size
+            self.num_attention_heads = getattr(self.config, "text_config", self.config).num_heads
+            self.embed_size_per_head = (
+                getattr(self.config, "text_config", self.config).hidden_size // self.num_attention_heads
+            )
+        else:
+            self.vocab_size = getattr(self.config, "decoder", self.config).vocab_size
+            self.num_attention_heads = getattr(self.config, "decoder", self.config).num_attention_heads
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_seq2seq.py:614]: does this mean that we don't need a normalized_config anymore ? (uniformization for  model's config parameters happened before the min supported transformers version)

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 21]
```diff
@@ -1088,89 +967,108 @@ def _from_pretrained(
             token=token,
             revision=revision,
         )
-
         if len(onnx_files) == 0:
             raise FileNotFoundError(f"Could not find any ONNX model file in {model_id}")
 
+        encoder_path = cls._infer_file_path(ENCODER_ONNX_FILE_PATTERN, encoder_file_name, onnx_files)
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_seq2seq.py]: question : do we have tests for this like https://github.com/huggingface/optimum-onnx/blob/main/tests/onnxruntime/test_decoder.py#L315 ?

**Review by echarlaix — APPROVED** [review]
Huuge work, thanks a lot @IlyasMoutawwakil

File: optimum/onnxruntime/modeling_decoder.py (hunk 19 above)
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_decoder.py]: yeah i think we can remove it

File: optimum/exporters/onnx/input_generators.py (hunk 15 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/input_generators.py]: yes should work as well !

File: optimum/exporters/onnx/model_configs.py (hunk 16 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/model_configs.py:2601]: ah yes this is still needed, the model is broken in transformers in that version, nice catch !

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 20 above)
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_seq2seq.py:614]: Yeah no need to use normalized configs, especially in modeling code where only very few key parameters are used, normalized configs create a dependency on optimum's normalized configs manager in modeling. Many contributors get overwhelmed by the fact they have to open two PRs in two different packages to add a model.
Another thing is that sometimes normalization in the way it is defined in optimum doesn't solve the issue entirely (see ORTForCausalLM where we still had to handle edge cases with multi_query, falcon's new architecture etc where num kv heads depends on those args as well).

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 21 above)
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_seq2seq.py]: that's what I added in https://github.com/huggingface/optimum-onnx/pull/34/commits/948fc48299370985628816cbaedddf521dae480a#diff-678ab0ec5ee9ec399ac21be989a3975f39962eff5176b839fd3759b05359aedeR593 (added test_load_model_from_hub and test_load_model_from_hub_infer_onnx_model)

Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_seq2seq.py]: I just pushed a [commit](https://github.com/huggingface/optimum-onnx/pull/34/commits/6bf36f2fec7afe2bc7e88e957e6e8f85f69ade70) that unifies all three modeling files around the use of `_infer_file_path` which is now documented in `ORTModel._infer_file_path`.
The simplest use case is in modeling_ort, I also used it in modeling_decoder to redefine the priority logic which fixed this edge case https://github.com/huggingface/optimum-onnx/pull/34/commits/6bf36f2fec7afe2bc7e88e957e6e8f85f69ade70#diff-7d70dee5f773f702f3f375d67dc8625febb7754852e84f61f4e7ca872b7cc1e2L416

Comment (echarlaix) [reply optimum/onnxruntime/modeling_seq2seq.py]: looks great, thanks a lot @IlyasMoutawwakil
